### ISSUE TITLE
NotSupportedError: DOM Exception 9: The implementation did not support the requested type of object or operation.

### ISSUE THREAD

**Issue by tobscure** [lhorie/mithril.js#917]
As of 270b20a2b029da7a5807648f43e31798eb7e0e96, I'm getting this error in Safari (9.0.1). Chrome is unaffected.

> NotSupportedError: DOM Exception 9: The implementation did not support the requested type of object or operation.

The stack trace points to `createContextualFragment` on [this line](https://github.com/lhorie/mithril.js/blob/270b20a2b029da7a5807648f43e31798eb7e0e96/mithril.js#L638). Looking back at the [previous commit](https://github.com/lhorie/mithril.js/blob/c1912d272648a43d1fedc2cdc1565a0634ee4f53/mithril.js#L1236), it looks like this part was refactored and a surrounding try/catch was removed?

**Comment by isiahmeadows** [comment]
PRs + tests welcome!

**Comment by nianwang** [comment]
Example on Safari 9.0.2 (11601.3.9), for tag `v0.2.2-rc1`

```
var example = {
  view: function() {
    return m('div', [
      m('p', '&copy;'),
      m('p', m.trust('&copy;')),
      m.trust('&copy;')
    ]);
  }
}

m.mount(document.body, example);
```

**Comment by isiahmeadows** [comment]
Is there a more minimal example?

**Comment by iamjohnlong** [comment]
Seem to be 
```
document.createRange().createContextualFragment('<div>I am a div node</div>')
```
Isn't supported in safari even though it has the native method. 

Would reverting back to a try/catch be acceptable @isiahmeadows ?

**Comment by isiahmeadows** [comment]
Yes, but pull the try/catch out into its own function to keep Chrome from refusing to optimize the whole function. What would be best would be to pull it out into its own independent function:

```js
function appendTextFragment(parentElement, data) {
  try {
    parentElement.appendChild($document.createRange().createContextualFragment(data))
  } catch (e) {
    parentElement.insertAdjacentHTML("beforeend", data)
  }
}

// And in injectHTML
if (nextSibling) {
  // ...
} else {
  appendTextFragment(parentElement, data)
}
```

**Comment by barneycarroll** [comment]
I was looking over Cito source earlier and noticed a lot of similar environment-dependent forking in frequently used functions. My ignorance of under-the-hood JIT optimisation logic leaves me somewhat unqualified, but wouldn't the best approach be to run a memoized function to test for this edge case on first execution and thereafter return the stored outcome?

Back in the day when I wrote engine-dependent implementation forks of repeatedly-invoked functions, the function reference itself would be the test, that upon execution re-assigned the reference to point to the engine-specific variation. I get the impression this completely short-circuits certain engines procedures for optimising run time, but never found a way to validate the assertion.

**Comment by iamjohnlong** [comment]
I'm not sure how to write a test for this @isiahmeadows 

https://github.com/lhorie/mithril.js/pull/923

**Comment by isiahmeadows** [comment]
@barneycarroll That won't work with `m.deps()`, since the mock will be used from the start. Otherwise, I would've already suggested that.

**Comment by isiahmeadows** [comment]
Maybe once we get rid of the stupid mock, *maybe* it might work.

**Comment by barneycarroll** [comment]
Argh. Yeah, I  see that problem ;)

**Comment by isiahmeadows** [comment]
The only thing that even *needs* mocked is the XMLHttpRequest, so we aren't actually requesting anything, and can still read headers.

**Comment by pygy** [comment]
Don't forget requestAnimationFrame for coercing async into sync behavior (unless you want to rewrite all the async tests on top of `m.hook`, if it ever lands...

**Comment by isiahmeadows** [comment]
@pygy That, too. I didn't think of that. But the rest is probably best suited to die in the fires of `/dev/null`. :smile:

**Comment by gabrieleceranto** [comment]
Hi all, i just encountered first time this bug. After some try, i found solution to let him work in safari (and other browsers too):
```javascript
var range = $document.createRange();
range.selectNode(parentElement);
parentElement.appendChild(range.createContextualFragment(data));
```

I originally found bug in safari 5.1.10, so i guess this bug affect all recent safari versions

**Comment by isiahmeadows** [comment]
@gabrieleceranto It's fixed: #923

**Comment by cmnstmntmn** [comment]
i still get it in Safari v. 9.0.2 (11601.3.9) with Mithril v0.2.2-rc.1

![screen shot 2016-02-03 at 15 47 36](https://cloud.githubusercontent.com/assets/2312755/12784046/7d02f30c-ca8d-11e5-96d6-d703bd22e1a7.png)

**Comment by tivac** [comment]
@cmnstmntmn Not surprising, `v0.2.2-rc1` was cut before the fix in #923 landed.

**Comment by isiahmeadows** [comment]
I think @lhorie will probably cut out another `-rc` at some point. There's a few other things remaining (like passing more of the new test suite).
